pypsa/statistics.py

```diff
@@ -396,24 +428,33 @@ def capex(
         @pass_empty_series_if_keyerror
         def func(n, c):
             col = n.df(c).eval(f"{nominal_attrs[c]}_opt * capital_cost")
-            if bus_carrier is not None:
-                masks = [
-                    port_mask(n, c, port=port, bus_carrier=bus_carrier)
-                    for port in get_ports(n, c)
-                ]
-                mask = reduce(np.logical_or, masks)
-                col = col[mask.astype(bool)]
-            return col
+            if bus_carrier is None:
+                grouper = get_grouper(n, c, groupby=groupby, nice_names=nice_names)
+                ds = col.groupby(grouper).agg(aggregate_groups)
+                return ds
+            else:
```

Comment:
I would leave the costs mostly untouched. Iterating and summing over different ports will lead to double counting of costs if more than one bus-carrier is selected. Think about a bus_carrier = ["gas", "electricity"], then the cost of components which are attached to both carriers at the same time are summed.